### PR TITLE
Integrate FFT in Batch Reconstruction

### PR DESCRIPTION
**Pull request by rahulgovind**

### CODE REVIEW

File: honeybadgermpc/batch_reconstruction.py
Comment (amiller) [inline honeybadgermpc/batch_reconstruction.py:97]: Good use of precomputed data here calling `fnt_decode_step1` only once for  batch

File: honeybadgermpc/polynomial.py
Comment (amiller) [inline honeybadgermpc/polynomial.py:369]: This `EvalPoint` helper looks like a good design, it's clear to see how it works and it encapsulates having to pass `omega` around separately from `point`

File: tests/test_batch_reconstruction.py
Comment (amiller) [inline tests/test_batch_reconstruction.py]: this test looks good, a bit repetitive versus `test()` but makes it clear that the evaluation is using the fft eval ponit

**Review by amiller — APPROVED** [review]
- really nice first PR
- this will probably get merged after #158 so you should pre-emptively make the new parts pep8 compliant, basically get rid of `camelCaseStyle` preferring `under_score_style` instead

**Comment by smkuls** [comment]
- You have some merge conflicts, can you please resolve those.

- There is a CI pipeline which runs after every PR, it looks like it didn't run for this PR - I think it is probably because there are conflicts - can you fix it and make sure the CI passes?

- I will review your code once the CI passes so please don't merge it until then.

File: honeybadgermpc/batch_reconstruction.py
Comment (smkuls) [inline honeybadgermpc/batch_reconstruction.py:101]: code style: I think there are certain guidelines around how you make the split to the next line..
I follow this
```
attempt_reconstruct(
chunk, field, n, t, point, precomputed_data)
```

The other option is to move each parameter to a separate line. What you have used is inconsistent with the rest of the code, can you please change this?

File: honeybadgermpc/batch_reconstruction.py
Comment (smkuls) [inline honeybadgermpc/batch_reconstruction.py:129]: same comment as above

File: honeybadgermpc/batch_reconstruction.py
Comment (smkuls) [inline honeybadgermpc/batch_reconstruction.py]: nit: Generally a bad practice to use the same variable name as the arg name `use_fft=use_fft`

File: tests/test_batch_reconstruction.py
Comment (rahulgovind) [reply tests/test_batch_reconstruction.py]: Should I separate out the code common to both tests?

File: honeybadgermpc/polynomial.py
Comment (smkuls) [inline honeybadgermpc/polynomial.py]: Please avoid using print statements anywhere. Use the logging framework instead. `logging.info` or `logging.debug` or `logging.warn`. Please take a pass over your code and get rid of all of them.

File: honeybadgermpc/polynomial.py
Comment (smkuls) [inline honeybadgermpc/polynomial.py]: what's `Pover`?

Comment (smkuls) [reply honeybadgermpc/polynomial.py]: Please remove any commented code.

File: honeybadgermpc/polynomial.py
Comment (smkuls) [inline honeybadgermpc/polynomial.py:381]: any reason why we want to fix the seed?

File: honeybadgermpc/polynomial.py
Comment (smkuls) [inline honeybadgermpc/polynomial.py:391]: nit: else is not needed since you have a `return`

File: honeybadgermpc/robust_reconstruction.py
Comment (smkuls) [inline honeybadgermpc/robust_reconstruction.py:26]: nit: consider renaming the arg value and name

File: honeybadgermpc/wb_interpolate.py
Comment (smkuls) [inline honeybadgermpc/wb_interpolate.py:147]: Same comment as above regarding splitting to next line

File: tests/test_batch_reconstruction.py
Comment (smkuls) [inline tests/test_batch_reconstruction.py]: Can you change it to `elliptic_curve.BLS...`? It is a larger modulus and we have it in most places in the existing code.

File: tests/test_batch_reconstruction.py
Comment (smkuls) [inline tests/test_batch_reconstruction.py]: remove commented code

File: tests/test_batch_reconstruction.py
Comment (smkuls) [inline tests/test_batch_reconstruction.py]: Same comment as above regarding splitting to next line

File: tests/test_batch_reconstruction.py
Comment (smkuls) [inline tests/test_batch_reconstruction.py]: Same comment as above regarding splitting to next line

File: tests/test_batch_reconstruction.py
Comment (smkuls) [inline tests/test_batch_reconstruction.py]: Do you think it will be better to separate these to multiple tests if they are separate? It shouldn't happen that the first test fails and then the rest of the test isn't executed (though Python might actually execute it)

File: tests/test_batch_reconstruction.py
Comment (smkuls) [inline tests/test_batch_reconstruction.py]: Same comment as above regarding splitting to next line

**Review by smkuls — CHANGES_REQUESTED** [review]

File: honeybadgermpc/polynomial.py
Comment (rahulgovind) [reply honeybadgermpc/polynomial.py]: Sure

File: honeybadgermpc/polynomial.py
Comment (rahulgovind) [reply honeybadgermpc/polynomial.py]: All right

File: honeybadgermpc/batch_reconstruction.py
Comment (rahulgovind) [reply honeybadgermpc/batch_reconstruction.py]: Why so?

File: honeybadgermpc/batch_reconstruction.py
Comment (rahulgovind) [reply honeybadgermpc/batch_reconstruction.py:129]: There's actually a lot of code (at least in batch_reconstruction.py and polynomial.py) that follows this style. I believe it'll actually be harder to just split the arguments to a different line if a function has a lot of arguments.

File: honeybadgermpc/polynomial.py
Comment (rahulgovind) [reply honeybadgermpc/polynomial.py:381]: All parties must arrive at the same omega for the algorithm to work. Fixing the seed was the easiest way to do that that I could think of.

File: honeybadgermpc/robust_reconstruction.py
Comment (rahulgovind) [reply honeybadgermpc/robust_reconstruction.py:26]: Are you referring to precomputed_data?

File: honeybadgermpc/batch_reconstruction.py
Comment (smkuls) [reply honeybadgermpc/batch_reconstruction.py:129]: I guess it is okay. You can resolve comments related to these.

I checked PEP8 - this seems to be a permitted style.

File: tests/test_batch_reconstruction.py
Comment (rahulgovind) [reply tests/test_batch_reconstruction.py]: Sure

File: honeybadgermpc/batch_reconstruction.py
Comment (smkuls) [reply honeybadgermpc/batch_reconstruction.py]: Andrew and I had a discussion the other day where we wanted to actually avoid code like `x = open(x)` and this reminded me of that - that would ease debugging and improve readability but I guess for function args those arguments don't hold.

I guess it is okay to leave it as is.

File: honeybadgermpc/polynomial.py
Comment (rahulgovind) [reply honeybadgermpc/polynomial.py:391]: That's true. But in this case, the 'else' statement makes it easier to figure out that the second route is taken only when fnt is not being used right?

File: honeybadgermpc/batch_reconstruction.py
Comment (rahulgovind) [reply honeybadgermpc/batch_reconstruction.py]: All right.

File: honeybadgermpc/polynomial.py
Comment (smkuls) [reply honeybadgermpc/polynomial.py:391]: It is way obvious with the return and there's only one statement in the if. `else` serves no purpose.

File: honeybadgermpc/robust_reconstruction.py
Comment (smkuls) [reply honeybadgermpc/robust_reconstruction.py:26]: I think this is fine for function args. Closing this.

File: tests/test_batch_reconstruction.py
Comment (rahulgovind) [reply tests/test_batch_reconstruction.py]: I think so too. That makes it easier to figure out what test is actually being run. But it would be better to set up common test cases inside by subclassing pytest.TestCase, since most of these tests have common setup code.

Comment (rahulgovind) [reply tests/test_batch_reconstruction.py]: For now, I just looked at how the other batch reconstruction tests were being done and used the same format.

Comment (smkuls) [reply tests/test_batch_reconstruction.py]: Yea the existing code is far from perfect. Most of it was written when we were in a time crunch and it has been reviewed properly or written properly.

File: honeybadgermpc/polynomial.py
Comment (rahulgovind) [reply honeybadgermpc/polynomial.py:381]: I'll document this in the code.

File: tests/test_batch_reconstruction.py
Comment (rahulgovind) [reply tests/test_batch_reconstruction.py]: Should I refactor the tests?

Comment (smkuls) [reply tests/test_batch_reconstruction.py]: Yea, it would be better to refactor them.

**Review by smkuls — APPROVED** [review]

**Pull request merged**
